crates/forge-runner/src/build_trace_data.rs

```diff
@@ -32,14 +32,20 @@ pub const TRACE_DIR: &str = ".snfoundry_trace";
 pub const TEST_CODE_CONTRACT_NAME: &str = "SNFORGE_TEST_CODE";
 pub const TEST_CODE_FUNCTION_NAME: &str = "SNFORGE_TEST_CODE_FUNCTION";
 
-#[must_use]
 pub fn build_profiler_call_trace(
     value: &Rc<RefCell<CallTrace>>,
     contracts_data: &ContractsData,
-) -> ProfilerCallTrace {
+) -> Option<ProfilerCallTrace> {
```

Comment:
We only need the `ProfilerCallTrace` when want to save the call trace data - that's how we determine if we want to collect the vm trace (since collecting the vm trace can impact the memory usage). We decide based on the `ExecutionDataToSave` if we need the vm trace or not. If we don't then all the `CallTrace` structs will contain `vm_trace: None`. Since profiler struct requires vm_trace we cannot build the profiler struct in that case (we don't need it anyways) - that's why I made the trace field in test result in `TestCaseSummary::<Single>::Passed` optional - so this function will return `None`.

I thought about it for a bit and imo it is the best option without introducing type families, which would change 1241241 signatures (really, tried it and gave up).

Reply:
The other idea was to pass `ExecutionData` to this function and assert that the `vm_trace` is indeed `Some` or `None` as implied by `ExecutionData`, but it adds even more params to functions and I didn't see a clear benefit of that other asserting that we for sure don't collect the trace when not needed - if we don't collect the trace when required the code will panic in `maybe_save_execution_data`

Reply:
Changed, no longer an option